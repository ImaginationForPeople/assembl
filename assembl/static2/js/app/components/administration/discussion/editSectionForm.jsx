// @flow
import React from 'react';
import { connect } from 'react-redux';
import { Translate, I18n } from 'react-redux-i18n';
import { FormGroup, Checkbox, OverlayTrigger, Button } from 'react-bootstrap';
import { getEntryValueForLocale } from '../../../utils/i18n';
import FormControlWithLabel from '../../common/formControlWithLabel';
import { deleteSectionTooltip, upTooltip, downTooltip } from '../../common/tooltips';
import {
  updateSectionTitle,
  updateSectionUrl,
  toggleExternalPage,
  deleteSection,
  moveSectionUp,
  moveSectionDown
} from '../../../actions/adminActions/adminSections';

type EditSectionFormProps = {
  url: string,
  type: string,
  title: string,
  nbSections: number,
  index: number,
  handleTitleChange: Function,
  handleUrlChange: Function,
  toggleExternalPageField: Function,
  handleDeleteClick: Function,
  handleDownClick: Function,
  handleUpClick: Function
};

const DumbEditSectionForm = ({
  url,
  type,
  title,
  nbSections,
  index,
  handleTitleChange,
  toggleExternalPageField,
  handleUrlChange,
  handleDeleteClick,
  handleDownClick,
  handleUpClick
}: EditSectionFormProps) => {
  const titlePh = I18n.t('administration.sections.titlePh');
  const urlPh = I18n.t('administration.sections.urlPh');
  const hasUrl = url !== null;
  return (
    <div className="form-container">
      <div className="title left">
        {`${index + 1}. `}
        {title || I18n.t('administration.sections.custom')}
      </div>
      <div className="right">
        {type !== 'HOMEPAGE' ? (
          <div className="inline">
            {index < nbSections - 1 ? (
              <OverlayTrigger placement="top" overlay={downTooltip}>
                <Button onClick={handleDownClick} className="admin-icons">
                  <span className="assembl-icon-down-bold grey" />
                </Button>
              </OverlayTrigger>
            ) : null}
            {index > 1 ? (
              <OverlayTrigger placement="top" overlay={upTooltip}>
                <Button onClick={handleUpClick} className="admin-icons">
                  <span className="assembl-icon-up-bold grey" />
                </Button>
              </OverlayTrigger>
            ) : null}
          </div>
        ) : null}
        {type === 'CUSTOM' ? (
          <OverlayTrigger placement="top" overlay={deleteSectionTooltip}>
            <Button onClick={handleDeleteClick} className="admin-icons">
              <span className="assembl-icon-delete grey" />
            </Button>
          </OverlayTrigger>
        ) : null}
      </div>
      <div className="clear" />
      {type !== 'HOMEPAGE' ? (
        <FormControlWithLabel label={titlePh} onChange={handleTitleChange} type="text" value={title} />
      ) : null}
      {type === 'HOMEPAGE' ? (
        <FormGroup>
          <Checkbox checked={hasUrl} onChange={toggleExternalPageField}>
            <Translate value="administration.sections.externalPage" style={{ lineHeight: '20px' }} />
          </Checkbox>
        </FormGroup>
      ) : null}
      {(type === 'HOMEPAGE' && hasUrl) || type === 'CUSTOM' ? (
        <FormControlWithLabel label={urlPh} onChange={handleUrlChange} type="text" value={url} />
      ) : null}
      <div className="separator" />
    </div>
  );
};

const mapStateToProps = (state, { id, editLocale }) => {
  const section = state.admin.sections.sectionsById.get(id);
  return {
    locale: state.i18n.locale, // for I18n.t()
    url: section.get('url'),
    type: section.get('type'),
    order: section.get('order'),
    title: getEntryValueForLocale(section.get('titleEntries'), editLocale, '')
  };
};

const mapDispatchToProps = (dispatch, { id, editLocale }) => ({
  handleTitleChange: e => dispatch(updateSectionTitle(id, editLocale, e.target.value)),
  handleUrlChange: e => dispatch(updateSectionUrl(id, e.target.value)),
  toggleExternalPageField: () => dispatch(toggleExternalPage(id)),
  handleDeleteClick: () => dispatch(deleteSection(id)),
  handleUpClick: () => dispatch(moveSectionUp(id)),
  handleDownClick: () => dispatch(moveSectionDown(id))
});

export { DumbEditSectionForm };

export default connect(mapStateToProps, mapDispatchToProps)(DumbEditSectionForm);