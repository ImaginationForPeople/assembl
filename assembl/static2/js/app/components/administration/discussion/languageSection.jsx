import React from 'react';
import { List } from 'immutable';
import { I18n, Translate } from 'react-redux-i18n';
import { compose, graphql } from 'react-apollo';
import { connect } from 'react-redux';
import { Checkbox } from 'react-bootstrap';

import SectionTitle from '../sectionTitle';
import withLoadingIndicator from '../../common/withLoadingIndicator';
import { addLanguagePreference, removeLanguagePreference, languagePreferencesHasChanged } from '../../../actions/adminActions';
import getAllPreferenceLanguage from '../../../graphql/AllLanguagePreferences.graphql';

class LanguageSection extends React.Component {
  constructor(props) {
    super(props);
    const { discussionLanguagePreferences, data } = props;

    const allLangs = {};
    data.defaultPreferences.languages.forEach((lang) => {
      allLangs[lang.locale] = { selected: false, name: lang.name };
    });

    discussionLanguagePreferences.forEach((locale) => {
      const prevState = allLangs[locale];
      allLangs[locale] = { ...prevState, selected: true };
    });

    this.state = {
      localeState: allLangs
    };
  }

  componentWillReceiveProps(nextProps) {
    // Manage the change in interface locale, causing language names to change
    const currentLocale = this.props.i18n.locale;
    const nextLocale = nextProps.i18n.locale;
    if (currentLocale !== nextLocale) {
      // Refresh cache with new information, as the translated names are returned from backend
      this.props.data.refetch({ variables: { inLocale: nextLocale } });
    }

    // Manage toggling of checkbox states from the store
    const totalLocaleList = List(Object.keys(this.state.localeState));
    const currentEditLocaleList = totalLocaleList.filter(locale => this.state.localeState[locale].selected).sort();
    const newLocalePreferences = nextProps.discussionLanguagePreferences.sort();
    // Only update the state if there is a change in language preferences
    if (!currentEditLocaleList.equals(newLocalePreferences)) {
      const newState = { ...this.state.localeState };
      Object.entries(newState).forEach(([locale, state]) => {
        const selectedState = state;
        if (newLocalePreferences.includes(locale)) {
          selectedState.selected = true;
        } else {
          selectedState.selected = false;
        }
      });
      this.setState({ localeState: newState });
    }
  }

  toggleLocale(locale) {
    const currentLocale = this.props.i18n.locale;
    this.props.signalLocaleChanged(true);
    const transientState = this.state.localeState[locale];
    const newState = { ...transientState, selected: !transientState.selected };
    if (newState.selected) {
      this.props.addLocaleToStore(locale);
    } else if (currentLocale !== locale) {
      this.props.removeLocaleFromStore(locale);
    }
  }

  render() {
    const currentLocale = this.props.i18n.locale;
    return (
      <div className="admin-box">
        <SectionTitle title={I18n.t('administration.discussion.0')} annotation={I18n.t('administration.annotation')} />
        <div className="admin-language-content">
          <div>
            <Translate value="administration.languageChoice" />
          </div>
          <form className="language-list">
            {Object.keys(this.state.localeState).map((locale) => {
              const localeData = this.state.localeState[locale];
              return (
                <Checkbox
                  checked={currentLocale === locale ? true : localeData.selected}
                  key={locale}
                  value={locale}
                  onChange={e => this.toggleLocale(e.target.value)}
                >
                  {localeData.name}
                </Checkbox>
              );
            })}
          </form>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ admin: { discussionLanguagePreferences }, i18n }) => ({
  i18n: i18n,
  discussionLanguagePreferences: discussionLanguagePreferences
});

const mapDispatchToProps = dispatch => ({
  addLocaleToStore: (locale) => {
    dispatch(addLanguagePreference(locale));
  },
  removeLocaleFromStore: (locale) => {
    dispatch(removeLanguagePreference(locale));
  },
  signalLocaleChanged: (state) => {
    dispatch(languagePreferencesHasChanged(state));
  }
});

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  graphql(getAllPreferenceLanguage, {
    options: props => ({
      variables: {
        inLocale: props.i18n.locale
      }
    })
  }),
  withLoadingIndicator()
)(LanguageSection);