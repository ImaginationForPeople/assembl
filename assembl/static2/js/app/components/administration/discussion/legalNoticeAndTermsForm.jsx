// @flow
import React from 'react';
import { Row } from 'react-bootstrap';
import { connect } from 'react-redux';
import { I18n } from 'react-redux-i18n';

import SectionTitle from '../sectionTitle';
import FormControlWithLabel from '../../common/formControlWithLabel';
import * as actions from '../../../actions/adminActions/legalNoticeAndTerms';
import type { State } from '../../../reducers/rootReducer';
import { getEntryValueForLocale } from '../../../utils/i18n';

type LegalNoticeAndTermsFormProps = {
  legalNotice: string,
  editLocale: string,
  termsAndConditions: string,
  updateLegalNotice: Function,
  updateTermsAndConditions: Function
};

export const DumbLegalNoticeAndTermsForm = ({
  legalNotice,
  editLocale,
  termsAndConditions,
  updateLegalNotice,
  updateTermsAndConditions
}: LegalNoticeAndTermsFormProps) => {
  const legalNoticeLabel = I18n.t('administration.legalNoticeAndTerms.legalNoticeLabel');
  const tacLabel = I18n.t('administration.legalNoticeAndTerms.termsAndConditionsLabel');
  return (
    <div className="admin-box">
      <SectionTitle title={I18n.t('administration.discussion.3')} annotation={I18n.t('administration.annotation')} />
      <div className="admin-content">
        <Row>
          <div className="form-container">
            <FormControlWithLabel
              key={`tac-${editLocale}-${termsAndConditions}`}
              label={`${tacLabel}*`}
              onChange={updateTermsAndConditions}
              required
              type="rich-text"
              value={termsAndConditions}
            />
            <div className="separator" />
            <FormControlWithLabel
              key={`legal-notice-${editLocale}-${legalNotice}`}
              label={`${legalNoticeLabel}*`}
              onChange={updateLegalNotice}
              required
              type="rich-text"
              value={legalNotice}
            />
          </div>
        </Row>
      </div>
    </div>
  );
};

const mapStateToProps = (state: State, { editLocale }: LegalNoticeAndTermsFormProps) => {
  const legalNoticeAndTerms = state.admin.legalNoticeAndTerms;
  const legalNotice = getEntryValueForLocale(legalNoticeAndTerms.get('legalNoticeEntries'), editLocale);
  const termsAndConditions = getEntryValueForLocale(legalNoticeAndTerms.get('termsAndConditionsEntries'), editLocale);
  return {
    legalNotice: legalNotice ? legalNotice.toJS() : '',
    termsAndConditions: termsAndConditions ? termsAndConditions.toJS() : ''
  };
};

const mapDispatchToProps = (dispatch: Function, { editLocale }: LegalNoticeAndTermsFormProps) => ({
  updateLegalNotice: (value: string) => dispatch(actions.updateLegalNoticeEntry(editLocale, value)),
  updateTermsAndConditions: (value: string) => dispatch(actions.updateTermsAndConditionsEntry(editLocale, value))
});

export default connect(mapStateToProps, mapDispatchToProps)(DumbLegalNoticeAndTermsForm);