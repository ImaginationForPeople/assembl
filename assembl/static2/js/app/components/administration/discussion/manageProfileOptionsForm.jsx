// @flow
import React from 'react';
import { Map, List } from 'immutable';
import { OverlayTrigger, Row, Button, Checkbox } from 'react-bootstrap';
import { connect } from 'react-redux';
import { I18n, Translate } from 'react-redux-i18n';

import SectionTitle from '../sectionTitle';
import TextField from './textField';
import { createRandomId } from '../../../utils/globalFunctions';
import { getEntryValueForLocale, type LangstringEntriesList } from '../../../utils/i18n';
import { addTextFieldTooltip, hideTextFieldTooltip } from '../../common/tooltips';
import * as actions from '../../../actions/adminActions/profileOptions';
import { displayModal, closeModal } from '../../../utils/utilityManager';

type TextFieldProps = {
  id: string,
  identifier: string,
  _toDelete: boolean,
  order: number,
  fieldType: string,
  titleEntries: LangstringEntriesList
};

type Props = {
  editLocale: string,
  textFields: List<Map<TextFieldProps>>,
  addSelectField: () => void,
  addTextField: () => void,
  deleteTextField: string => void,
  moveTextFieldDown: () => void,
  moveTextFieldUp: () => void,
  toggleTextFieldRequired: string => void,
  toggleTextFieldHidden: string => void,
  updateTextFieldTitle: (string, string, string) => void
};

const ManageProfileOptionsForm = ({
  addSelectField,
  addTextField,
  deleteTextField,
  editLocale,
  moveTextFieldDown,
  moveTextFieldUp,
  textFields,
  toggleTextFieldRequired,
  toggleTextFieldHidden,
  updateTextFieldTitle
}: Props) => {
  const confirmTextFieldDeletion = (id: string) => {
    const modalTitle = <Translate value="administration.confirmTextFieldDeletionTitle" />;
    const body = <Translate value="administration.confirmTextFieldDeletion" />;
    const footer = [
      <Button key="cancel" onClick={closeModal} className="button-cancel button-dark">
        <Translate value="debate.confirmDeletionButtonCancel" />
      </Button>,
      <Button
        key="delete"
        onClick={() => {
          deleteTextField(id);
        }}
        className="button-submit button-dark"
      >
        <Translate value="debate.confirmDeletionButtonDelete" />
      </Button>
    ];
    const includeFooter = true;
    return displayModal(modalTitle, body, includeFooter, footer);
  };

  const showAddFieldModal = () => {
    // const modalTitle = <Translate value="administration.profileOptions.addTextField" />;
    const modalTitle = '';
    // const body = <Translate value="administration.confirmTextFieldDeletion" />;
    const body = (
      <div>
        <Translate value="administration.profileOptions.createNewFieldModalBody" />
      </div>
    );
    const footer = [
      <Button
        key="textField"
        className="button-cancel button-generic"
        onClick={() => {
          closeModal();
          addTextField();
        }}
      >
        <Translate value="administration.profileOptions.choiceTextField" />
      </Button>,
      <Button
        key="selectField"
        className="button-cancel button-generic"
        onClick={() => {
          closeModal();
          addSelectField();
        }}
      >
        <Translate value="administration.profileOptions.choiceSelectField" />
      </Button>
    ];
    const includeFooter = true;
    return displayModal(modalTitle, body, includeFooter, footer);
  };

  const isNotHideable = (textField) => {
    switch (textField.get('identifier')) {
    case 'FULLNAME':
    case 'EMAIL':
    case 'PASSWORD':
    case 'PASSWORD2':
      return true;
    default:
      return false;
    }
  };

  return (
    <div className="admin-box">
      <SectionTitle title={I18n.t('administration.discussion.3')} annotation={I18n.t('administration.annotation')} />
      <div className="intro-text">
        <Translate value="administration.profileOptions.introText" />
      </div>
      <div className="admin-content">
        <Row>
          <div className="form-container profile-options">
            <form>
              {textFields.map((tf, idx) => (
                <div className="flex" key={tf.get('id')}>
                  <OverlayTrigger placement="top" overlay={hideTextFieldTooltip}>
                    <Checkbox className="textfield-checkbox" checked={tf.get('hidden')} disabled={isNotHideable(tf)} onChange={() => toggleTextFieldHidden(tf.get('id'))} />
                  </OverlayTrigger>
                  <TextField
                    deleteField={() => confirmTextFieldDeletion(tf.get('id'))}
                    fieldType={tf.get('fieldType')}
                    id={tf.get('id')}
                    identifier={tf.get('identifier')}
                    isFirst={idx === 0}
                    isLast={idx === textFields.size - 1}
                    moveDown={moveTextFieldDown}
                    moveUp={moveTextFieldUp}
                    required={tf.get('required')}
                    title={getEntryValueForLocale(tf.get('titleEntries'), editLocale, '') || ''}
                    toggleRequired={() => {
                      toggleTextFieldRequired(tf.get('id'));
                    }}
                    updateTitle={value => updateTextFieldTitle(tf.get('id'), editLocale, value)}
                    isSelectField={!!tf.get('options')}
                  />
                </div>
              ))}
              <OverlayTrigger placement="top" overlay={addTextFieldTooltip}>
                <div onClick={showAddFieldModal} className="plus margin-l">
                  +
                </div>
              </OverlayTrigger>
            </form>
          </div>
        </Row>
      </div>
    </div>
  );
};

const mapStateToProps = ({
  admin: {
    editLocale,
    profileOptions: { textFieldsById }
  }
}) => ({
  editLocale: editLocale,
  textFields: textFieldsById
    .filter(tf => !tf.get('_toDelete'))
    .sortBy(tf => tf.get('order'))
    .toList()
});

const mapDispatchToProps = dispatch => ({
  addSelectField: () => {
    const fieldId = createRandomId();
    dispatch(actions.addTextField(fieldId, 'select'));
    dispatch(actions.addSelectFieldOption(fieldId, createRandomId()));
    dispatch(actions.addSelectFieldOption(fieldId, createRandomId()));
    dispatch(actions.addSelectFieldOption(fieldId, createRandomId()));
  },
  addTextField: () => dispatch(actions.addTextField(createRandomId(), 'text')),
  deleteTextField: (id) => {
    closeModal();
    dispatch(actions.deleteTextField(id));
  },
  updateTextFieldTitle: (id, locale, value) => dispatch(actions.updateTextFieldTitle(id, locale, value)),
  toggleTextFieldRequired: id => dispatch(actions.toggleTextFieldRequired(id)),
  toggleTextFieldHidden: id => dispatch(actions.toggleTextFieldHidden(id)),
  moveTextFieldDown: id => dispatch(actions.moveTextFieldDown(id)),
  moveTextFieldUp: id => dispatch(actions.moveTextFieldUp(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ManageProfileOptionsForm);