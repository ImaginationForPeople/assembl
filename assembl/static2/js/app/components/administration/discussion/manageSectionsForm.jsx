// @flow
import React from 'react';
import { connect } from 'react-redux';
import { I18n } from 'react-redux-i18n';
import { OverlayTrigger } from 'react-bootstrap';
import { List } from 'immutable';

import SectionTitle from '../sectionTitle';
import EditSectionForm from './editSectionForm';
import { addSectionTooltip } from '../../common/tooltips';
import * as actions from '../../../actions/adminActions/adminSections';

type ManageSectionFormProps = {
  sections: List<string>,
  editLocale: string,
  createSection: Function
};

const DumbManageSectionsForm = ({ sections, editLocale, createSection }: ManageSectionFormProps) => (
  <div className="admin-box">
    <SectionTitle title={I18n.t('administration.sections.sectionsTitle')} annotation={I18n.t('administration.annotation')} />
    <div className="admin-content">
      <form>
        {sections.map((id, index) => (
          <EditSectionForm key={id} id={id} index={index} editLocale={editLocale} nbSections={sections.size} />
        ))}
        <OverlayTrigger placement="top" overlay={addSectionTooltip}>
          <div onClick={() => createSection(sections.size)} className="plus margin-l">
            +
          </div>
        </OverlayTrigger>
      </form>
    </div>
  </div>
);

const mapStateToProps = (state) => {
  const { sectionsInOrder } = state.admin.sections;
  return {
    editLocale: state.admin.editLocale,
    locale: state.i18n.locale, // for I18n.t()
    sections: sectionsInOrder
  };
};

const mapDispatchToProps = dispatch => ({
  createSection: (nextOrder) => {
    const newId = Math.round(Math.random() * -1000000).toString();
    return dispatch(actions.createSection(newId, nextOrder));
  }
});

export { DumbManageSectionsForm };

export default connect(mapStateToProps, mapDispatchToProps)(DumbManageSectionsForm);