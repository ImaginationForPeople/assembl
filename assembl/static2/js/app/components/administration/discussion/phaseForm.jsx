// @flow
import React from 'react';
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import { connect } from 'react-redux';
import { Translate, I18n } from 'react-redux-i18n';
import { updateStartDate, updateEndDate } from '../../../actions/adminActions/timeline';

type PhaseFormProps = {
  phaseNumber: number,
  start: moment$Moment,
  end: moment$Moment,
  handleStartDateChange: Function,
  handleEndDateChange: Function,
  locale: string,
  hasConflictingDates: boolean
};

export const DumbPhaseForm = ({
  phaseNumber,
  handleStartDateChange,
  handleEndDateChange,
  start,
  end,
  hasConflictingDates,
  locale
}: PhaseFormProps) => {
  const startDatePickerPlaceholder = I18n.t('administration.timelineAdmin.selectStart', { count: phaseNumber });
  const endDatePickerPlaceholder = I18n.t('administration.timelineAdmin.selectEnd', { count: phaseNumber });

  return (
    <div className="phase-form">
      <div className="date-picker-field">
        <div className="date-picker-type">
          <Translate value="search.datefilter.from" />
        </div>
        <label htmlFor="start-datepicker" className="datepicker-label">
          <DatePicker
            placeholderText={startDatePickerPlaceholder}
            selected={start}
            id="start-datepicker"
            onChange={handleStartDateChange}
            showTimeSelect
            timeFormat="HH:mm"
            dateFormat="LLL"
            locale={locale}
            shouldCloseOnSelect
            className={hasConflictingDates ? 'warning' : ''}
          />
          <div className="icon-schedule-container">
            <span className="assembl-icon-schedule grey" />
          </div>
        </label>
        {hasConflictingDates && (
          <div className="warning-label">
            <Translate value="administration.timelineAdmin.warningLabel" />
          </div>
        )}
      </div>
      <div className="date-picker-field">
        <div className="date-picker-type">
          <Translate value="search.datefilter.to" />
        </div>
        <label htmlFor="end-datepicker" className="datepicker-label">
          <DatePicker
            placeholderText={endDatePickerPlaceholder}
            id="end-datepicker"
            selected={end}
            onChange={handleEndDateChange}
            showTimeSelect
            timeFormat="HH:mm"
            dateFormat="LLL"
            locale={locale}
            shouldCloseOnSelect
            className={hasConflictingDates ? 'warning' : ''}
          />
          <div className="icon-schedule-container">
            <span className="assembl-icon-schedule grey" />
          </div>
        </label>
        {hasConflictingDates && (
          <div className="warning-label">
            <Translate value="administration.timelineAdmin.warningLabel" />
          </div>
        )}
      </div>
    </div>
  );
};

const mapStateToProps = (state, { phaseId }) => {
  const phase = state.admin.timeline.phasesById.get(phaseId);
  return {
    identifier: phase && phase.get('identifier'),
    start: phase ? phase.get('start') : null,
    end: phase ? phase.get('end') : null,
    hasConflictingDates: phase ? phase.get('hasConflictingDates') : null,
    locale: state.i18n.locale
  };
};

const mapDispatchToProps = (dispatch, { phaseId }) => ({
  handleStartDateChange: date => dispatch(updateStartDate(phaseId, date)),
  handleEndDateChange: date => dispatch(updateEndDate(phaseId, date))
});

export default connect(mapStateToProps, mapDispatchToProps)(DumbPhaseForm);