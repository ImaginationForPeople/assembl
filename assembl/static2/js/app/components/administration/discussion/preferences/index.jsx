// @flow
import React from 'react';
import { connect } from 'react-redux';
import { type ApolloClient, compose, withApollo } from 'react-apollo';
import { I18n, Translate } from 'react-redux-i18n';
import arrayMutators from 'final-form-arrays';
import { Field, FormSpy } from 'react-final-form';
import classnames from 'classnames';
import Loader from '../../../common/loader';
import LoadSaveReinitializeForm from '../../../form/LoadSaveReinitializeForm';
import AdminForm from '../../../../components/form/adminForm';
import CheckboxFieldAdapter from '../../../../components/form/checkboxFieldAdapter';
import CheckboxListFieldAdapter from '../../../form/checkboxListFieldAdapter';
import SectionTitle from '../../../../components/administration/sectionTitle';
import { load, postLoadFormat } from './load';
import { save, createMutationsPromises } from './save';
import validate from './validate';
import TextFieldAdapter from '../../../form/textFieldAdapter';
import { slugAllowedCharacters } from '../../../../constants';
import OnChangeFieldSetField from './OnChangeFieldSetField';
import type { DiscussionPreferencesFormValues } from './types.flow';

type Props = {
  client: ApolloClient,
  locale: string
};

type State = {
  slugIsInvalid: boolean,
  translationServiceMissing: boolean
};

type FormState = {
  values: DiscussionPreferencesFormValues
};

const loading = <Loader />;

class DiscussionPreferencesForm extends React.Component<Props, State> {
  state = {
    slugIsInvalid: false,
    translationServiceMissing: false
  };

  checkSlugValidity = (slug: ?string) => {
    if (slug) {
      this.setState({ slugIsInvalid: !slug.match(slugAllowedCharacters) });
    }
  };

  render() {
    const { client, locale } = this.props;
    const { slugIsInvalid, translationServiceMissing } = this.state;
    return (
      <div className="discussion-admin admin-box admin-content">
        <SectionTitle title={I18n.t('administration.menu.preferences')} annotation="" />
        <LoadSaveReinitializeForm
          load={(fetchPolicy: FetchPolicy) => load(client, fetchPolicy, locale)}
          loading={loading}
          postLoadFormat={postLoadFormat}
          createMutationsPromises={createMutationsPromises(client)}
          save={save}
          validate={validate}
          withWarningModal
          warningValues={['slug']}
          warningMessageKey="administration.slugWarning"
          mutators={{
            ...arrayMutators
          }}
          render={({ handleSubmit, pristine, submitting, values }) => (
            <div className="admin-content">
              <AdminForm handleSubmit={handleSubmit} pristine={pristine} submitting={submitting} disableSave={slugIsInvalid}>
                <FormSpy
                  subscription={{ values: true }}
                  // $FlowFixMe ignore undefined state
                  onChange={(state: FormState) => {
                    const { values: { withSemanticAnalysis, withTranslation } } = state;
                    this.setState({ translationServiceMissing: withSemanticAnalysis && !withTranslation });
                  }}
                />
                <div className="form-container">
                  <div className="title">
                    <Translate value="administration.informationOfTheDebate" />
                  </div>
                  <Field
                    validate={this.checkSlugValidity}
                    component={TextFieldAdapter}
                    name="slug"
                    label={I18n.t('administration.discussionSlug')}
                    className={classnames({ warning: slugIsInvalid })}
                  />
                  {slugIsInvalid ? (
                    <div className="warning-label">
                      <Translate value="administration.invalidSlug" />
                    </div>
                  ) : null}
                  <div className="separator" />
                  <div className="title">
                    <Translate value="administration.languageChoice" />
                  </div>
                  <div className="margin-l" />
                  <Field component={CheckboxListFieldAdapter} name="languages" />
                  <div className="separator" />
                  <div className="title">
                    <Translate value="administration.moderation" />
                  </div>
                  <Field
                    component={CheckboxFieldAdapter}
                    name="withModeration"
                    isChecked={values.withModeration}
                    label={I18n.t('administration.activateModeration')}
                    type="checkbox"
                  />
                  <div className="separator" />
                  <div className="title">
                    <Translate value="administration.translationService.title" />
                  </div>
                  {translationServiceMissing ? (
                    <div className="warning-label">
                      <Translate value="administration.translationService.description" />
                    </div>
                  ) : null}
                  <Field
                    component={CheckboxFieldAdapter}
                    name="withTranslation"
                    isChecked
                    label={I18n.t('administration.translationService.activate')}
                    type="checkbox"
                  />
                  <div className="separator" />
                  <OnChangeFieldSetField field="withSemanticAnalysis" becomes set="withTranslation" to />
                  <div className="title">
                    <Translate value="administration.semanticAnalysis.title" />
                  </div>
                  <div className="label-indication">
                    <Translate value="administration.semanticAnalysis.description" />
                  </div>
                  <Field
                    component={CheckboxFieldAdapter}
                    name="withSemanticAnalysis"
                    isChecked
                    label={I18n.t('administration.semanticAnalysis.activate')}
                    type="checkbox"
                  />
                </div>
              </AdminForm>
            </div>
          )}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  locale: state.i18n.locale
});

export { DiscussionPreferencesForm };

export default compose(connect(mapStateToProps), withApollo)(DiscussionPreferencesForm);