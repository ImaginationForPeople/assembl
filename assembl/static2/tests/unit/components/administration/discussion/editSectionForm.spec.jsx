import React from 'react';
import renderer from 'react-test-renderer';

import { DumbEditSectionForm } from '../../../../../js/app/components/administration/discussion/editSectionForm';
import '../../../../helpers/setupTranslations';

describe('DumbEditSectionForm component', () => {
  it('should render a form to edit a section of Assembl', () => {
    const handleTitleChangeSpy = jest.fn(() => {});
    const handleUrlChangeSpy = jest.fn(() => {});
    const toggleExternalPageFieldSpy = jest.fn(() => {});
    const handleDeleteClickSpy = jest.fn(() => {});
    const handleDownClickSpy = jest.fn(() => {});
    const handleUpClickSpy = jest.fn(() => {});
    const props = {
      handleTitleChange: handleTitleChangeSpy,
      handleUrlChange: handleUrlChangeSpy,
      toggleExternalPageField: toggleExternalPageFieldSpy,
      handleDeleteClick: handleDeleteClickSpy,
      handleDownClick: handleDownClickSpy,
      handleUpClick: handleUpClickSpy,
      url: 'www.google.fr',
      type: 'HOMEPAGE',
      title: 'Accueil',
      nbSections: 3,
      index: 0
    };
    const component = renderer.create(<DumbEditSectionForm {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});