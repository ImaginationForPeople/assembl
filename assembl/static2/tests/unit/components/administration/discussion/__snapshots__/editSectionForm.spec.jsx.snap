// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DumbEditSectionForm component should render a form to edit a section of Assembl 1`] = `
<div
  className="form-container"
>
  <div
    className="title left"
  >
    1. 
    Accueil
  </div>
  <div
    className="right"
  />
  <div
    className="clear"
  />
  <div
    className="form-group"
  >
    <div
      className="checkbox"
      style={undefined}
    >
      <label
        title=""
      >
        <input
          checked={true}
          disabled={false}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className={undefined}
          style={
            Object {
              "lineHeight": "20px",
            }
          }
        >
          Use external page
        </span>
      </label>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor={undefined}
    >
      URL
    </label>
    <input
      className="form-control"
      disabled={undefined}
      id={undefined}
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="URL"
      type="text"
      value="www.google.fr"
    />
  </div>
  <div
    className="separator"
  />
</div>
`;
